src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -2149,6 +2159,111 @@ static void assertBindIdentifierTargets(InMethodBinder? inMethodBinder, Concurre
+                    case TupleElementSyntax tupleElement when tupleElement.Identifier == token:
+                    case FromClauseSyntax fromClause when fromClause.Identifier == token:
+                    case LetClauseSyntax letClause when letClause.Identifier == token:
+                    case JoinClauseSyntax joinClause when joinClause.Identifier == token:
+                    case JoinIntoClauseSyntax joinIntoClause when joinIntoClause.Identifier == token:
+                    case QueryContinuationSyntax queryContinuation when queryContinuation.Identifier == token:
+                    case LocalFunctionStatementSyntax localFunctionStatement when localFunctionStatement.Identifier == token:
+                    case VariableDeclaratorSyntax variableDeclarator when variableDeclarator.Identifier == token:
+                    case SingleVariableDesignationSyntax singleVariable when singleVariable.Identifier == token:
+                    case LabeledStatementSyntax labeledStatement when labeledStatement.Identifier == token:
+                    case ForEachStatementSyntax forEachStatement when forEachStatement.Identifier == token:
+                    case CatchDeclarationSyntax catchDeclaration when catchDeclaration.Identifier == token:
+                    case TypeParameterSyntax typeParameter when typeParameter.Identifier == token:
+                    case ParameterSyntax parameter when parameter.Identifier == token:
+                        switch (token.Text)
+                        {
+                            case "field":
+                                if (method.AssociatedSymbol is PropertySymbol { IsIndexer: false })
+                                {
+                                    // Report "field" conflict with keyword.
+                                    reportConflict(syntax, token, diagnostics);
+                                }
+                                break;
+                            case "value":
+                                if (method.MethodKind == MethodKind.PropertySet)
+                                {
+                                    // Record the potential "value" conflict, and report conflicts later,
+                                    // after dropping any that refer to the implicit parameter.
+                                    valueIdentifiers ??= PooledDictionary<SyntaxNode, SyntaxToken>.GetInstance();
+                                    valueIdentifiers.Add(syntax, token);
+                                }
+                                break;
+                        }
+                        break;
+                    default:
+                        // If we reach here, add the unhandled identifier case to the
+                        // switch cases above (and test in FieldAndValueKeywordTests).
+                        Debug.Assert(false, $"Unhandled identifier: parent {syntax.Kind()}, token {token}");
```

Comment:
>Debug.Assert [](http://example.com/codeflow?start=24&length=12)

It is questionable how much we can rely on this asset since it is behind a language version check.